### ISSUE TITLE
[feature request] Swarm mode should support batch/cron jobs in addition to persistent services

### ISSUE THREAD

**Issue by nathanleclaire** [moby/moby#23880]
## Description

Today Swarm mode allows users to specify a group of homogenous containers which are meant to be kept running with the `docker service` CLI.  This abstraction, while powerful, may not be the right fit for containers which are _intended_ to eventually terminate or only run periodically.

Consider, for instance:
- An admin who wishes to allow users to submit long-running compiler jobs on a Swarm cluster
- A website which needs to process all user uploaded images into thumbnails of various sizes
- An operator who wishes to periodically run `docker rmi $(docker images --filter dangling=true -q)` on each machine
## Problem

Though some use cases could potentially be implemented by a service which pulls jobs off a distributed queue, there are some issues with this approach:
1. It adds an operational burden of creating, administrating, and ensuring the health of such a queue.  For many, this will kick up the barrier to entry of performing such tasks.
2. It does not necessarily ensure that failed jobs can be re-run the correct number of times or with the correct parameters.  Like above, this burden has now been offloaded to the user instead of being natively orchestrated.
3. It does not allow for easy customization of job parallelism settings.

Anything which is intended to be run periodically (such as the image garbage collection example above), could potentially cause a thundering herd problem if the scheduling is not handled by Swarm.  Imagine, for instance, that a user creates a `docker service` to periodically run a command inside the containers using normal old `cron`.  If these all wake up and attempt to execute at the same time during a time when production traffic is surging, it may be an issue for more critical production web services.  While it may help to mitigate the problem if the users do proper capacity planning and use flags such as `--reserve-cpu`, separating these concerns early on, especially when a temporal element is involved, seems to prudent.  (thanks to @jpetazzo who originally pointed out these concerns to me, and probably will have good insight as well)

This issue requests and outlines a proposed CLI, to get the ball rolling on discussion, track the issue, and gather information about potential use cases.
## Proposal

A new top-level command, `docker jobs`, could be introduced.  It would allow to specify that a container should be run X times, or every Y interval of time.  It could also be used to check up on these jobs.

Examples.

Run batch job once:

``` console
$ docker jobs create \
    -e S3_INPUT_BUCKET \
    -e S3_OUTPUT_BUCKET \
    -e IMAGE_NAME \
    nathanleclaire/convert
3x3bq2ibh1qe

$ docker jobs wait 3x3bq2ibh1qe; docker jobs ls
ID              NEXT PREV           FINISHED   FAILURES IMAGE
3x3bq2ibh1qe    -    3 seconds ago  1/1        0/1      nathanleclaire/convert 
```

Run batch job 16 times:

```
$ docker jobs create \
    --runs 16 \
    --parallel 3 \
    nathanleclaire/failer
bku4f1s1ncm0

$ docker jobs ls
ID            NEXT  PREV           FINISHED   FAILURES IMAGE
3x3bq2ibh1qe  -     2 minutes ago  1/1        0/1      nathanleclaire/convert 
bku4f1s1ncm0  Now   Now            4/16       6/10     nathanleclaire/failer
```

Run a task every hour:

```
$ docker jobs create \
    --every 1hr \
    nathanleclaire/hourlytask
ddh7pqvgbd8l

$ # One hour later...

$ docker jobs ls
ID            NEXT PREV         FINISHED  FAILURES IMAGE
3x3bq2ibh1qe  -    1 hour ago   1/1       0/1      nathanleclaire/convert 
bku4f1s1ncm0  -    1 hour ago   16/16     6/20     nathanleclaire/failer 
ddh7pqvgbd8l  1hr  1 minute ago 1/1       0/1      nathanleclaire/hourlytask
```

Interactively re-run a job:

``` console
$ docker jobs restart 3x3bq2ibh1qe
Running job 3x3bq2ibh1qe again from the beginning
```

Alternatively, `service` model could be expanded to accommodate this?  But it seems they would be easier to manage (for users) as separate things.

Please let me know what you think (when you get a chance -- please focus on 1.12 first and foremost ;) ) @aluzzardi @vieux @stevvooe @abronan and others.
## Cute animal

Since I'm requesting feature, least I can do is provide cute animal picture.

![image](https://cloud.githubusercontent.com/assets/1476820/16287318/23914b16-3897-11e6-99e8-5b92f84ec517.png)

**Comment by AkihiroSuda** [comment]
My "indexed job" suggestion seems related to and can be integrated to this suggestion.

https://github.com/docker/docker/issues/23843
```
$ echo "apple banana cherry" > fruits.txt
$ echo "green yellow red" > colors.txt
$ docker jobs create --parallel 3
     --scatter=FRUIT="$(cat fruits.txt)" \
     --scatter=COLOR="$(cat colors.txt)" \
     busybox
     'echo "Have a nice $COLOR $FRUIT" && sleep 5'
(container0) Have a nice green apple
(container1) Have a nice yellow banana
(container2) Have a nice red cherry
```

**Comment by gerred** [comment]
What if the Tasks API were expanded as its own fully featured toolset, allowing individual tasks to be submitted and scheduled separate of services? Task as it stands could expand "ServiceAnnotations" to be more, generally "Annotations". Jobs and services are already abstractions upon the tasks, but it could be an easy way to allow arbitrary task controllers to handle different use cases, while including the basic abstractions of service and job in the engine itself.

This would also necessitate having more knowledge of the dispatcher that these controllers could draw from, so my suggestion might be a little pie in the sky at this point.

**Comment by kelseyhightower** [comment]
@gerred I think you are on to something as there are other systems that work the way you describe. You can have different controllers process jobs and monitor them for successful completion. All controllers can reuse a single scheduler to ensure proper workload placement, but it would be the individual controllers responsibility to ensure long running services are always restarted or rescheduled in face of node failures. In the case of jobs a Job Controller can ensure tasks run at least once until successful completion.

**Comment by aluzzardi** [comment]
@nathanleclaire @kelseyhightower @gerred: The use case is already covered by the service model.

Services have a "mode" (I don't like this term, very open to suggestions) and a mode maps directly to an orchestrator (what @kelseyhightower refers to as a controller).

We currently have two service modes: Replicated ("scalable") and Global ("runs on every machine").

e.g. `docker service create --mode global redis`

Every mode has its own options and attributes, so in the future we could achieve this with something like:
```
$ docker service create \
    --mode batch
    --every 1hr \
    nathanleclaire/hourlytask
ddh7pqvgbd8l
```

/cc @dongluochen @aaronlehmann

**Comment by treksler** [comment]
First off, batch mode for docker service would be great. There would need to be a way to easily and intuitively list batch jobs, their schedules, when a job runs next, when it ran last, was it successful, etc.

I suppose `docker service ls -f "mode=batch"` would work, but it's not really as intuitive as `docker job ls`

Similarly, `docker service tasks -f "mode=batch"` could show last execution status in LAST STATE and next scheduled run in DESIRED STATE or maybe just show "Scheduled" or "Batch" in DESIRED STATE and have` docker service inspect` display the batch schedule

The proposed `--parallel` is the same thing as `--replicas`, as far as I can see.
i.e. how many copies of this should be spawned?

The proposed `--runs` is somewhat conceptually similar to `--restart-max-attempts`
i.e. how many times do you want to (attempt to) run it, after it exits?

It looks like you can already kind of fake batch mode with `--restart-delay` set to something like 24 hours to create a pseudo daily job, but if I need to make sure my container/job runs exactly 30 seconds from midnight on some swarm node every day, then I am hooped without a batch mode.

Something like this might work
`docker service create -f "mode=batch" --batch-run-at="23:59:30" someimage:1.0`
or we could support cron syntax
`docker service create -f "mode=batch" --batch-run-schedule="59 23 * * * " someimage:1.0`

The problem with commands that take a --mode parameter is that only a subset of command line switches will be applicable to each mode. At that point, they may as well be separate commands.
It would be hard to tell which [command line switches](https://docs.docker.com/engine/reference/commandline/service_create/) on `docker service create` apply to which mode. This is why I added the --batch-run prefix to the switches in the above examples.

**Comment by aleksejlopasov** [comment]
I have the same problem with my services that must run periodically. I've created a service which runs a container with a php script inside of it. This script must run every day at the same time. But when the script finishes his work, container stops and that's all. I can't run it again from the master-node. the only way is to remove the service and recreate it. So docker swarm mode doesn't fit the situations when you run the services that are expected to exit

**Comment by nathanleclaire** [comment]
> I have the same problem with my services that must run periodically. I've created a service which runs a container with a php script inside of it. This script must run every day at the same time. But when the script finishes his work, container stops and that's all. I can't run it again from the master-node. the only way is to remove the service and recreate it. So docker swarm mode doesn't fit the situations when you run the services that are expected to exit

FWIW, you might be able to work around this in the interim by running cron inside of the container, so that it wakes up and runs the script same time every day and container doesn't exit -- e.g. https://forums.docker.com/t/running-cronjob-in-debian-jessie-container/17527/5

**Comment by stevvooe** [comment]
> So docker swarm mode doesn't fit the situations when you run the services that are expected to exit.

At this time, swarm mode is focused on long-running services.

To be honest, in a production system, I have been unhappy nearly every time I've used cron. The solution provided by @nathanleclaire is great but the cron approach has problems when you _need_ something to run, particularly when you want to customize deadline behavior and backfill. There are also issues involving how cron stores state to disk that can be problematic if the container is rescheduled elsewhere.

This is merely a personal opinion and doesn't mean we won't implement batch in the future, cause we will.

The best way to handle this, I've found, is to have the script do its own scheduling. Here is the simplest version the runs a script every minute:

```
#!/bin/sh

sleep 60
exec myjob.sh
```

While this is a toy, it has a few nice properties:

1. Every time it exits, it will be scheduled to a free node.
2. Simple as dirt.

The main issue is that each time it is run, it waits 60 seconds. That means if you lose the node, the task will get rescheduled and wait 60 seconds before running. This is a huge problem when we replace that with 24 hours or 2 days.

We can solve this with something like anacron, but that typically means the cron data to do backfill is located on a single node. 

So, what do we do? Basically, to solve this, we want to calculate the wait time each time we run and ensure that the state is saved to commonly accessible component in the cluster. The answer to most software engineering problems is to use Redis!

The script is surprisingly simple:

```bash
#!/bin/sh

set -e

# query the ttl and set to wait, use --csv so we don't have to parse output
wait=$(redis-cli -h $REDIS_HOST -p $REDIS_PORT --csv  ttl $SCRIPT_KEY)

case $wait in 
  -2) # redis code for key does not exist
     myjob.sh
     # on success of our job, we set the ttl to the environment variable provided in the service
     # make sure the script has set -e
     redis-cli set $SCRIPT_KEY "$(date)" EX $INTERVAL_SECONDS
    fi
     ;;
   *)
    sleep $wait # sleep until the key expires
    # we can also have this sleep some modulus of wait time or have 
    # a maximum to ensure the task is rescheduled to other nodes periodically.
    ;;
esac
```

So, with this, we set `REDIS_HOST`, `REDIS_PORT`, `SCRIPT_KEY` and `INTERVAL_SECONDS`, we can get a script to run every X seconds that will be resilient to failures and rescheduling on nodes.

This isn't too bad for a small amount of effort. Obviously, with some effort, this can get a lot more sophisticated. An enterprising solution good make a version of this script that could simply take an environment variable with the cron schedule and do the right thing. The key here, being two properties:

1. Enable the script to schedule itself.
2. Store the run state in a distributed manner (some database).

An even more ambitious version could bind mount the docker socket right into the task and run a container for the script, separating the scheduler and scripts that are being scheduled.

Obviously, this isn't ideal, but this is a solid and flexible hack.

Note that this doesn't mean we won't implement batch.

**Comment by aleksejlopasov** [comment]
To @nathanleclaire and @stevvooe
I've found another solution:
What if I check the state of the service before running it with 'docker service create'? If it is completed that I will delete it and create new service. If not - than echo message 'bla-bla, sorry, the service is running at the moment, try again later'.

**Comment by nathanleclaire** [comment]
> I've found another solution:
> What if I check the state of the service before running it with 'docker service create'? If it is 
> completed that I will delete it and create new service. If not - than echo message 'bla bla,sorry, 
> the service is running at the moment, try again later'.

Why use a `docker service` at all in this case?  `docker service` today is intended for processes that you want to keep running.  There will never be a "stopped service" like you mention.  Have a play with `docker service` to get a feel for how this works.

```console
$ docker service create --name deadservice alpine sh -c 'sleep 1; echo See ya later!'
aen8savf90lqy1i9i842wy498

$ docker service ps deadservice
ID                         NAME               IMAGE   NODE  DESIRED STATE  CURRENT STATE        ERROR
cydq1r3d2piw4z6b779r39say  deadservice.1      alpine  moby  Ready          Ready 2 days ago
7zqj89296nke3158ci3rv13zw   \_ deadservice.1  alpine  moby  Shutdown       Complete 2 days ago
3c678vy2zdi9ew1otxn909u0r   \_ deadservice.1  alpine  moby  Shutdown       Complete 2 days ago
a2xo7bq99w8kcx5a8csgc4uh7   \_ deadservice.1  alpine  moby  Shutdown       Complete 2 days ago
```

Your solution just depends on how complex your needs are.  If the job is safe to run more/less often than intended (or possibly concurrently) then you could probably get by with something like a `docker service` that just does `while sleep 60; do ./command.sh; done`.  If it's not safe to run willy-nilly then you have to put more elbow grease into rigging up some details to try and ensure safe / consistent operation.  But doing so is one of those things which looks simple at first glance and turns out to be fiendishly difficult in practice.  It's why tech like Apache Kafka exists to try and give us more guarantees about message delivery and event ordering.

**Comment by aleksejlopasov** [comment]
@nathanleclaire 
Maybe I had to provide you with more detailes. I run 'docker service create' with '--restart-condition none' paramether. There is a php script inside the container, that is run by the service. When it's done, the container on the worker-node will stop and the service on the master-node will stay created but the state will change to 'complete'. I can't run that service again, but I can check its state and run it again with 'docker service create' as I told in previous post

**Comment by rayjohnson** [comment]
We use rundeck to talk to a docker swarm to run "batch" jobs.  At enterprise level you need a lot more logic around notifications, specify the ways you do retries, changing dependent jobs, etc.  While a docker specific solution would be cool - it strikes me it should be a separate service than extending docker swarm in weird ways...

**Comment by rogaha** [comment]
@nathanleclaire have you thought about job dependencies? e.g. 
```
docker jobs create \
    --every 1hr \
    nathanleclaire/hourlytask
bku4f1s1ncm0
docker jobs create \
   --depends-on bku4f1s1ncm0
    rogaha/aggregationJob
```

I think it would be a nice addition to your proposal.

**Comment by stevvooe** [comment]
> have you thought about job dependencies?

@rogaha And this is where the scope creep begins ;). There are a hundred ways to represent dependency graphs, both temporal and resource-based, so getting that right makes things challenging.

I would rather we provide the functionality to build cron systems on top of swarm services. Something like this:

```
docker service create --name mycronservice -e EVERY=1hr stevvooe/dockacron stevvooe/myhourlyscript
```

Effectively, this would run an image that runs a scheduler `stevvooe/dockacron` that runs the container `stevvooe/myhourlyscript` every hour. This requires the following features:

1. Services that can create new tasks with some sort of plugin API.
2. Some sort of persistent storage.

I think this can be done today by bind mounting the docker socket and constraining the service to manager nodes, while providing volumes for the state storage.

**Comment by rayjohnson** [comment]
When compared to Rundeck, ActiveBatch or any other enterprise level batch management systems this proposal will never be of much use.  Retry mechanisms, dependent jobs, triggered jobs, and many other features that this would never compete with.  Alerting, job history and a full "run book" view are also typically critical features.  It is scope creep to jump into this area and it would not add a lot of value.

What would add value, however, is to make it easy for an external batch scheduling system to call into swarm and run a job.  The old swarm did this with "docker run" fairly well.  The new service command doesn't allow for this so well.  I'd suggest focusing more on enabling existing scheduling systems to use swarm over trying to recreate the wheel in yet another area.

**Comment by stevvooe** [comment]
@rayjohnson I'm not sure that comparing with such systems is an apt comparison. The tools you are referencing are UIs and dependency management on top of some sort of execution system. From your response, I also sense some confusion about the use cases that services and swarm standalone hit.

From what you've filed in https://github.com/docker/swarmkit/issues/1501, I'd recommend you continue using standalone swarm with Docker 1.12, rather than the built in swarm mode features.

However we extend swarm-mode will be a likely target for more feature rich job scheduling systems. You're right: the service command isn't designed for batch. I've touched on some the requirements in https://github.com/docker/docker/issues/23880#issuecomment-237343613, but your points about scope creep are a consideration. There are a million ways to schedule interdependent batch jobs and this is largely why this isn't in the first version of Docker services.

**Comment by mcallaway** [comment]
The comments about "job dependencies" and the implied "feature creep" makes me want to point out that there is a small jump between dependent jobs and full blown "workflows". This is what I'm interested in. It's probably too big for this issue, but I wanted to just put this out there so that other eyes might see it and maybe lead someplace else.

https://github.com/common-workflow-language/common-workflow-language

There's a need for workflow execution that leverages docker for the runtime environment for each "node" in a DAG (for example). CWL has arisen out of this need. But there's a lack of implementations. It's related to this issue, but I realize bigger. Maybe someone can link to other more relevant connections?

**Comment by stevvooe** [comment]
@mcallaway There are several ways of doing workflow specifications in different, spanning various levels of sophistication and complexity. CWL is a great example, but there are a number spanning the hadoop ecosystem to more academic applications. Many companies just grow their own to fit their specific needs, as well.

Any batch features in docker would like be positioned to support or complement these kinds of systems, rather define a competing workflow system. The challenge to this project is cutting out the right primitives to make that painless.

**Comment by OferE** [comment]
+1 - very important.

**Comment by yank1** [comment]
+1

**Comment by danielwegener** [comment]
Maybe you guys could also help me understand the wording a bit or correct me if I got something wrong:
So _docker standalone swarm_ (now legacy?) supported the `docker run` api (used by pretty much every third party tool interacting with docker). Now the  _new docker swarm mode_ (since 0.12.x, enabling _docker swarmkit_) brings support for awesome services but drops support for docker swarm standalone.

If I got that right, just a curious question/idea: How about _simply_ (I bet it's not?) reenabling the _legacy_ docker client api for the docker swarm such that both, _swarmkits_ `docker service` but also the _swarm standalone_ `docker run` command could be executed on the same swarm?
If that does not make so much sense (since docker standalone swarm seems to be an _external project_), how about adding support for `docker run` (with standalone swarm semantic) to swarmkit?

**Comment by ehazlett** [comment]
@danielwegener we are working towards parity with "docker run" (see https://github.com/docker/docker/issues/25303 for related) as well as consolidating the API to support what you are asking for.  i'm not sure on time table as swarmkit is quite different in its model vs. docker run.

**Comment by ehazlett** [comment]
@danielwegener i should also add, while not ideal, you can run swarm classic on a swarm mode enabled cluster to give you both.

**Comment by rayjohnson** [comment]
We are running both and can attest that the approach does indeed work.  However, I can also agree that it is less than ideal - basically you have to manage two clusters in one and thus have twice the administrative issues.  :)

**Comment by danielwegener** [comment]
@ehazlett @rayjohnson  Thanks for the quick response and clarification. I'll give the stacked cluster a try :)

**Comment by chrisguidry** [comment]
While it's not perfect, I did have some measure of success with @aleksejlopasov's suggestion to use  `--restart-condition none`.  Here's how it works:

Let's say I have a container image whose `ENTRYPOINT` is a command that runs for 15 seconds and then exits normally.

```
$ docker service create --name my_scheduled_task --replicas 1 --restart-condition none ....
```

Then using a normal `cron` task:
```
* * * * * root docker service update my_scheduled_task
```

This causes the Swarm to, every minute, schedule an instance of the container on one of the swarm nodes, allowing it to run to completion.  Also, if the task happens to take too long during one or more runs, you won't end up with stacked up `cron`s because Swarm is already that it has enough replicas.

Some very important caveats:

* If the process exits with an error code, Swarm will immediately try to reschedule it.
* You won't get any error reporting or failure notifications from `cron` about this (since the command `docker service update my_scheduled_task` will return quickly and won't be attached to your container)


I have only used this in a few toy examples, and it seems to work well for those.  I think you'd want to have another error reporting mechanism ready because it would be very easy to miss a failing task.

**Comment by lsl** [comment]
Trying to find a solution I like for this problem. 

Using services (@chrisguidry), might work you could do a replication reduction to 0 then to 1 if you want it rescheduled on the best node (which is kind of the point) - you'd need something to reduce replicas to 0 after completion with an external test, either that or do it straight away with a really high timeout for shutdown.

Fwiw / reference, k8s has a decent implementation which is pretty much what you want:
https://kubernetes.io/docs/user-guide/cron-jobs/

**Comment by chrisguidry** [comment]
Yeah, k8s has the notion of just a "Job" which is probably what that was built on.

The more time I spend with Swarm mode, the more I love it (it's far better than the classic Swarm in every important respect).  Adding a simple one-shot job (as opposed to a long-running service) would just complete this picture for me.

**Comment by sedouard** [comment]
@chrisguidry I'm assuming you run those cron scheduled tasks on just one master? What if that master goes down? How has it been working out?

**Comment by chrisguidry** [comment]
@sedouard What I do is duplicate the cron tasks on multiple servers (I think I do all the odd-numbered ones in my cluster).  This might be somewhat hacky, but the idea is that `docker service update my_scheduled_task` seems to be an idempotent operation.  If the service is currently running, it doesn't do anything.  So as long as the jobs take a little bit of time and the clocks are relatively in sync, you get the effect of a distributed redundancy cron job and it being a singleton.

**Comment by chrisguidry** [comment]
Again, my method is clearly a temporary hack until we get `Job`s ;)

**Comment by taiidani** [comment]
While we're in sharing mode and waiting _eagerly_ for native functionality, I set mine up with Jenkins using runners on each of my masters and periodic builds. The runners have the docker socket mounted to them, and extract the tag of their corresponding web service to determine which version of the image to use.

```
# Given a service called rates_php using an image from company/rates, build a .env file
docker service inspect --format '{{.Spec.TaskTemplate.ContainerSpec.Image}}' rates_php | sed 's/company\/rates:/TAG=/' > .env

# The docker-compose definition has `image: vacasa/rates:$TAG` in it, which it will get from the .env file
docker-compose run --rm --name=rates-cron php php artisan rates:listings
```

This lets me run crons on any of the masters.

Drawbacks:
* All the networks the crons access must be `attachable`
* Since the masters are the only instances that can query the service API (and are the only ones guaranteed to have the overlay networks I need on them), the crons _must_ be run on the masters.
* This method obviously spikes CPU/Mem/disk usage on the most important machines in the cluster, which terrifies me.

**Comment by rayjohnson** [comment]
I like the hack - but it would not work for us.  Mainly because we use Rundeck and both want to send out alerts on job failure and want to control how many times a retry occurs before it stops trying all together.

That's why I'm kind of worried if the future solution is to try and recreate job control or cron in swarm as opposed to have a simple way to run a single job ones on the cluster and then finish.  Not that it wouldn't be useful for many use cases.  But enterprise level job schedulers sure have a lot of functionality that swarm will probably never have.

We now have many dozens of jobs running like this with the old swarm.  We are finding we are using many of those advanced features.  All I need from the new swarm is to run a single container and know when it finishes.  To me that would be way more flexible and powerful to then combine with something like RunDeck then having a simplistic cron like job scheduler.

**Comment by taiidani** [comment]
Hmm, @rayjohnson not sure if you're talking about chrisguidry or my solution -- while mine is far from perfect, Jenkins does give us log output in the Jenkins UI, success/failure tracking, DataDog & Slack integration, etc.

All those benefits are coming from the Jenkins side though and how it lets us create slaved-out runner containers on the masters. I'm not sure if Rundeck would have the same capabilities.

**Comment by rayjohnson** [comment]
Ha - you post snuck in before I got mine submitted!  Actually, we considered and played with using Jenkins as the scheduling system.  Depending not he types of jobs Jenkins might be better.  For the jobs we have RunDeck was better suited.  Same idea though - RunDeck is just the scheduler which then spawns the job in the swarm cluster.  Logs come back to rundeck, metrics on job running times & success, etc.  I think we could integrate with slack but we don't do that.

**Comment by bbdimitriu** [comment]
Did anyone try the integration of RunDeck with the new Swarm? I'd really prefer to avoid deploying the two Swarms together... I'd imagine one still needs to use @aleksejlopasov 's "hack" this way for now anyway.

**Comment by lsl** [comment]
@chrisguidry its early but this looks like a good temporary solution until we get something native - https://github.com/alexellis/jaas

log support requires --experimental and looks a little buggy but it does the trick

Is it me or is this feature just not that important for most people? What do most people do? 

Roll their own scheduler? Use patterns that don't require scheduling?  

I've got a few ways to tackle this with the daemon route but I still think docker should be doing it because it can do resource balancing properly.

**Comment by ms1111** [comment]
@lsl jaas looks promising. For non-critical scheduled tasks I'm just recreating a service from cron with `--restart-policy=none` on one of the masters. However the Swarm scheduler will still occasionally restart the service on its own for no apparent reason - this was happening after elections, last time I looked into it.

**Comment by aluzzardi** [comment]
/cc @aaronlehmann

"None" should not restart on elections, probably a bug

**Comment by aaronlehmann** [comment]
> /cc @aaronlehmann
> 
> "None" should not restart on elections, probably a bug

Known issue, I think. Remember that restart condition is not considered in service reconcilliation, only the steady state. A leader election triggers this reconciliation.

In other words, if you set up a service with 5 replicas, one exits, and then there's a leader election, the new leader will see that there are only 4 replicas running and start a new one.

I agree we should find a way to fix this, but so far it has been unclear how to keep track of this state (5 replicas not meaning 5 replicas). In the past I've thought about using the dead task objects in the reconciliation process, but this would not be very robust.

There's a swarmkit issue for this, I think.

**Comment by AkihiroSuda** [comment]
@aaronlehmann Do we have a github issue ticket for that issue?

**Comment by aaronlehmann** [comment]
> @aaronlehmann Do we have a github issue ticket for that issue?
> 

https://github.com/docker/swarmkit/issues/932

**Comment by xiusir** [comment]
@rogaha 
can't agree more!

**Comment by nasskach** [comment]
Hi,

Is this feature will be taken in consideration or do we need to wait one more year ? Seriously, this feature is already available under k8s and AWS ECS, what are we waiting for?

Thanks.

**Comment by lsl** [comment]
How many people would be satisfied with a new deploy option "cron" / "schedule". e.g. in compose form:

```
services:
  my-worker:
    image: lsl/my-worker
    deploy:
      replicas: 10
      cron: "0 */6 * * *"
```

I suspect this would be relatively easy to implement with the restart_policy.delay logic that already exists.

**Comment by rayjohnson** [comment]
Man - all I need is the ability to run one container to completion on the swarm.  I do need to be able to get logs and know how the job ends.  If we have that we can use cron, RunDeck, or anything else.  We do not need swarm to implement cron to do cron jobs - we need it to run a single freakin job without going through freakin hoops to know it completed and not get restarted.

Ray

On Jun 14, 2017, at 9:17 PM, Louis Laugesen <notifications@github.com<mailto:notifications@github.com>> wrote:


How many people would be satisfied with a new deploy option "cron" / "schedule". e.g. in compose form:

services:
  my-worker:
    image: lsl/my-worker
    deploy:
      replicas: 10
      cron: "0 */6 * * *"


I suspect this would be relatively easy to implement with the restart_policy.delay logic that already exists.

-

**Comment by Vanuan** [comment]
@rayjohnson I think you already can. Just use restart policy "never".

**Comment by ms1111** [comment]
@Vanuan Yes, just keep in mind that services with restart policy "never" tend to get restarted randomly because of docker/swarmkit#932. So this is only safe for certain types of job.

**Comment by lsl** [comment]
So seems like we have two distinct use cases here.

- Cron-like service scheduling
- Docker run for swarm mode (& persistent logs)

@ms1111 @Vanuan finding similar issues using long restart delays to solve the cron-like scheduling problem.

**Comment by taybin** [comment]
Kubernetes supports these higher level concepts:

https://kubernetes.io/docs/concepts/workloads/controllers/jobs-run-to-completion/
https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/

It'd be great to see support for these sort of workloads at the compose file level.

**Comment by spatialvlad** [comment]
Cron-like functionality would definitely be helpful.

It looks like the discussion in this topic found a way to run a service once using restart condition being none.

I guess I risk sounding obvious, but I might anyway mention it any, as someone might find it helpful.

It looks like it is also possible to have a basic repetitive repetition of running a job by using the delay option in the restart policy. E.g. to run a script daily (e.g. for a daily backup) this options would work:
```
            restart_policy:
                condition: any
                delay: 24h
```

Modifying the delay time one can get some other repetitive cycles, like weekly, etc.
It does not really give good control of the exact time of when the service will run. But I guess this can be somewhat "hacked" now by starting or updating the service at needed time.

Anyway, seeing native support for cron-like functionality would be great.

**Comment by taiidani** [comment]
Well our solution of using docker-compose run against the stack file is running out of steam. I am having a hard time coming to grips with Docker's strategy of adding first-class Secrets and Config objects and then expecting us (I assume) to find alternatives for scheduled tasks and one-offs.

What is Docker's recommendation for handling Secrets in cron jobs? Should I abandon these features in favor of Hashicorp Vault, or implement workarounds like the restart condition option above where gathering logs and exit codes is a chore?

**Comment by Vanuan** [comment]
Wouldn't this work:

1. Add the one-off to docker-compose.yml
2. run docker deploy
3. remove it
4. run docker deploy again

**Comment by shouze** [comment]
@Vanuan yup, but for cron jobs that won't work. Nice trick BTW.

**Comment by Vanuan** [comment]
Yeah, for cron jobs you'd have to either use restart delay (it wouldn't be at the same time though) or emded cron into the container itself.

**Comment by taiidani** [comment]
There are also the issues of:
* Blocking your script while you wait for the one-off service to exit so you don't kill it accidentally midway-through. Now that we have Swarm events that may work too but you'd have to register and listen for the task exit event.
* Navigating the service API to get the exit code of the task so you can determine if it failed/succeeded.
* Navigating the service API (assuming you have access to `service logs`) to gather the logs, or giving your cron runner a message along the lines of "Logs aren't supported. Check your log aggregator please! ¯\\_(ツ)_/¯"

At this point I'm coming to terms with having to tell our engineers, "You get insecure crons, or you get Secrets. You can't have both" while we continue our `docker-compose run` strategy.

**Comment by Vanuan** [comment]
Well, I don't see how Docker prevents solving those issues. Those aren't unique to the swarm mode.

I imagine some supervisor script which will run your one-off script and push completion status to some database/message queue, while your dependent service waits for completion status to appear.

Similarly for gathering logs, I don't see what are you expecting. Docker doesn't prevent you from accessing logs. You can mount them to some network drive to be accessible wherever you want them to be.

I don't see how this is related to secrets at all.

**Comment by Vanuan** [comment]
Reading it more carefully, I think I understood.

You've exploited docker-compose run to run some script in the foreground, i.e. "synchronous" commands.
E.g.:

```
docker-compose run mysql mysql -e "INSERT ..."
docker-compose run mysql mysql -e "CREATE ..."
```

That wasn't designed for that use case.

**Comment by Jota87** [comment]
Hey, I was struggling with this problem too. My solution has been add an environment tag on my docker compose for my swarm:

> curator:
> image: bobrik/curator:latest
> deploy:
> mode: replicated
> replicas: 1
> secrets:
> - curator_yml
> - actionfile_yml
> command: --config /run/secrets/curator_yml /run/secrets/actionfile_yml
> environment:
> - SCHEDULE='* * * * *'

**Comment by mide** [comment]
@Jota87, I think that the `SCHEDULE` environment variable you mentioned has been implemented  at the image level, not at the orchestration level. One main difference: I'd wager that when you run that curator image, it _stays running_. If it were a real cron-like task, it'd do the work and go away.

I'll happily be wrong because this means the feature is closer than I previously thought 😄 .

**Comment by Jota87** [comment]
you are right, this doesn't works! is trying to rerun curator all the time and burning my machines 😄😄 don't do that! i was exited with the idea but not has been my best 😄

**Comment by shouze** [comment]
yup, one solution if none managed by docker swarm orchestrator itself would be to have a tool like [traefik](https://github.com/containous/traefik) and pass sheduling through sercive labels that would be listened by this tool, that would run only on swarm manager nodes of course.

**Comment by getvivekv** [comment]
I am currently working on a proof of concept for this requirement, the theory is that the one-off cron jobs container can be implemented by creating a cronjob container in swarm mode that can create new swarm service in the cluster and once the service is finished running the cronjob, it just stops (restart condition: none). Having a container that has access to the host node's docker socket, can create `docker service` in the cluster with one replica and that's whole purpose is to start the container, execute the cronjob (cronjobs are fed via an environment variable) and die.

**Comment by stevelacy** [comment]
This is how we ended up using swarm mode to run "one off" services:

```go
    // command is from a network request
    serviceSpec := swarm.ServiceSpec{
      TaskTemplate: swarm.TaskSpec{
        ContainerSpec: &swarm.ContainerSpec{
          Image: "alpine",
          Command: command,
          StopSignal: "SIGINT",
        },

        RestartPolicy: &swarm.RestartPolicy{
          Condition: "none",
        },
      },
    }
    resp, err := cli.ServiceCreate(ctx, serviceSpec, types.ServiceCreateOptions{})
```
That code is running inside a container on a swarm master node as a container, not service. It then listens and connects to volume `/var/run/docker/sock:/var/run/docker.sock` and allows the docker SDK to manage new services as above.

**Comment by getvivekv** [comment]
Yes, exactly what I said above. I am still working on customizing the logic for our application. Glad to see that it works.

**Comment by stevelacy** [comment]
@getvictor Check this out:
```bash
docker service create --name quartermaster \
  -e TOKEN=4jrs8-534js-345ds-3lrd0 \
  -p 9090:9090 \
  --constraint 'node.role == manager' \
  --mount type=bind,source=/var/run/docker.sock,destination=/var/run/docker.sock \
  stevelacy/quartermaster
```
Confirmed that it works, specified `node.role == manager` to ensure it has manager access to the swarm.

**Comment by stevelacy** [comment]
@getvivekv Check this out:
```bash
docker service create --name quartermaster \
  -e TOKEN=4jrs8-534js-345ds-3lrd0 \
  -p 9090:9090 \
  --constraint 'node.role == manager' \
  --mount type=bind,source=/var/run/docker.sock,destination=/var/run/docker.sock \
  stevelacy/quartermaster
```
Confirmed that it works, specified `node.role == manager` to ensure it has manager access to the swarm.

**Comment by shouze** [comment]
ok @stevelacy related to this github repo I guess https://github.com/stevelacy/quartermaster will give a try, looks pretty good as a basis

**Comment by zsluedem** [comment]
i think one-shot job for more situations then cron job. Sometimes i just want to deploy a one time job and not a cron job based on my app.One-shot job is more flexible instead of cron one~!

**Comment by n3tw0rk5** [comment]
What about repeatable one jobs that vary with each run eg:
docker run --rm -ti -v $PWD:$PWD -w $PWD image:tag /bin/bash -c “python script with an argument that changes for each container run & an input file that changes as well”

How would this work in a swarm where you may have multiple job submissions, so you want to scale?

**Comment by nasskach** [comment]
@getvivekv yes we did something similar using [rundeck](http://rundeck.org/) packaged in a container running on manager nodes only.

**Comment by getvivekv** [comment]
Assuming that rundeck needs to be installed on a server, how do you take care of the cron if rundeck server goes down ? Are you starting a swarm service with rundeck? @nasskach

**Comment by nasskach** [comment]
@getvivekv As we have many manager nodes in our cluster, if the node where rundeck is deployed goes down, it container will be normally moved to another manager node.

I'm currently thinking* about writing a small cron application that will run also in a container on a manager node but that will read cron tasks from the labels of the different containers... like traefik is doing for load balancing.

I'm new at Golang programming and I'm still not sure that it is a good idea ^^

**Comment by getvivekv** [comment]
Yes my question was about rundeck server going down, in this case you don't have redundancy. You can't initiate the cron job. We are internally discussing if we can use AWS lambda to initiate the cronjob

**Comment by Gowiem** [comment]
For folks coming here looking for a non hacky solution right now, one possibility is to go the route of something like [Clockwork](https://github.com/Rykian/clockwork) or [python schedule](https://github.com/dbader/schedule) or the equivalent for your language. Then you can have a long running service which kicks off jobs on the schedule you determine within that system.

**Comment by pageflex** [comment]
We are looking to move our backoffice support  to task based and to use docker to save infrastructure costs. The ability to stand a container up as an image either on a schedule or from an external scheduler is a necessary feature. We'd love to have both those features with swarm for the redundancy, and the ability to also have long standing services within the same infrastructure would be a nice value add proposition. 

I see all the hacks and we'll have to go that route but I'm curious if these ideas have been placed to a roadmap somewhere that can give use an expectation of when this thread will be addressed.

**Comment by s17t** [comment]
I have a couple of use cases that would fit perfectly in one-off job feature. Now I'll have to just round robin on N docker server. Even without the cron part this feature is more than desirable; there are plenty of schedulers out there.

Here [0] someone wrote a wrapper just for having one-off job at swarm level after the 1.12. 

[0] https://github.com/alexellis/jaas

**Comment by alexellis** [comment]
@s17t thanks for mentioning JaaS - if anyone would like to get jobs on Swarm this is probably one of the simplest ways you can do it right now. 

Contributions are welcome - https://github.com/alexellis/jaas

OpenFaaS is also related on this topic for shorter running tasks - seconds to minutes - https://github.com/openfaas/faas

Would be ideal to see this in Swarm but we've been waiting two years for an update.

**Comment by bitsofinfo** [comment]
+1

**Comment by BretFisher** [comment]
For those looking, I'm listing some projects that might help solve this in addition to [Alex's JaaS](https://github.com/alexellis/jaas) (most mentioned above as well):

https://github.com/rayyanqcri/swarm-scheduler
https://github.com/vfarcic/docker-flow-cron
https://github.com/stevelacy/quartermaster
https://www.rundeck.com/blog/introduction-to-docker-plugins-for-rundeck

**Comment by borNfreee** [comment]
In addition to @BretFisher list

* https://github.com/opsxcq/tasker

**Comment by cpuguy83** [comment]
What are all the properties a job might have?

(mostly taken from @nathanleclaire)

1. Quantity of tasks to run
2. Control of task parallelism
3. period/frequency of runs
i. How might this affect quantity/parallelism?
ii. Do "cron" style periodic jobs fit into the same model as one-off jobs? i.e. does it make sense to have a service mode called "job" and the job spec defines a period/frequency (or not) or should there be two different modes such  as "batch" and "periodic"?
4. Deadlines

What else belongs here?

Do batch jobs need to know (in the container) some job number? Would this be a requirement or a nice to have (which could open up more cases down the line)?

What happens on periodic jobs if a job is not completed yet and it's time to run another? (maybe this is the parallelism tie-in)

I agree with @aluzzardi that this should be tied to the service model which is well designed for encapsulating different modes of operation of some workload.

**Comment by cpuguy83** [comment]
And just read stevvooe's posts as well on cron, I think I agree there. If we can enable the ability for a long-running service to manage periodic tasks without pulling them in as a top-level feature, this is a huge win.

**Comment by crazy-max** [comment]
Maybe this could help : https://github.com/crazy-max/swarm-cronjob

**Comment by djmaze** [comment]
@crazy-max That looks strikingly similar to the solution I was imagining to build myself. Awesome!

**Comment by arseniybanayev** [comment]
> When compared to Rundeck, ActiveBatch or any other enterprise level batch management systems this proposal will never be of much use. Retry mechanisms, dependent jobs, triggered jobs, and many other features that this would never compete with. Alerting, job history and a full "run book" view are also typically critical features. It is scope creep to jump into this area and it would not add a lot of value.
> 
> What would add value, however, is to make it easy for an external batch scheduling system to call into swarm and run a job. The old swarm did this with "docker run" fairly well. The new service command doesn't allow for this so well. I'd suggest focusing more on enabling existing scheduling systems to use swarm over trying to recreate the wheel in yet another area.

Completely agreed. Two prominent examples of existing scheduling systems trying to use swarm in a `docker run` fashion are:
1. Jenkins plugin to spin up single-use Jenkins agents in a swarm https://github.com/jenkinsci/docker-swarm-plugin
2. Airflow operator (in contrib) for running Airflow DAG tasks in "ephemeral" services https://github.com/apache/airflow/blob/master/airflow/contrib/operators/docker_swarm_operator.py

I would suggest that we improve the APIs with e.g. a command to bundle creating a service, waiting for its task(s) to complete, and removing the service. Let Airflow, Jenkins et al solve the problems of scheduling.

**Comment by cpuguy83** [comment]
Good news, a job api is on the way: https://github.com/moby/moby/pull/40307

**Comment by AkihiroSuda** [comment]
merged https://github.com/moby/moby/pull/40307
